_posts/2024/2024-02-26-validate-email-addresses.md

```diff
@@ -0,0 +1,134 @@
+---
+title: "Validate Email Addresses"
+description: "There’s a surprising number of ways to do it"
+layout: article
+category: ruby
+date: 2024-02-26 06:00
+image:
+  base: "2024/validate-email-addresses"
+  alt: "Envelopes"
+  credit: "Joanna Kosinska"
+  source: "https://unsplash.com/photos/envelope-paper-lot-uGcDWKN91Fs"
+---
+
+Ensuring you have data that makes sense is why Rails provides [validations](https://guides.rubyonrails.org/active_record_validations.html) through the Active Model library that underpins Active Record.
+
+Making sure you can email your users is the most important thing to get right in your application, so you probably already have validation checks around your `User#email` attribute to make your communications will get through.
+
+## Instead of…
+
+…inventing your own regular expression, or using this one from the [older Rails docs](https://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates).
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i },
+    presence: true,
+    uniqueness: { case_insensitive: true }
+end
+```
+
+## Use…
+
+…one of several decent options:
+
+### Devise
+
+Which might be a good choice, if you’re already using it.
+
+```ruby
+class User < ApplicationRecord
+  validates :email,
+    format: { with: Devise.email_regexp },
+    presence: true,
+    uniqueness: { case_insensitive: true }
+end
+```
+
+### `URI`
+
+There has been a regular expression built into the `URI` of Ruby‘s standard library since version 2.2. Here's the [commit adding it](https://github.com/ruby/ruby/commit/e63ab5d3ad289767eab49787e4e33390b0ce74e1)
```

Comment:
```suggestion
There has been a regular expression built into the `URI` of Ruby‘s standard library since version 2.2. Here's the [commit adding it](https://github.com/ruby/ruby/commit/e63ab5d3ad289767eab49787e4e33390b0ce74e1).
```